src/python.cpp

```diff
@@ -978,14 +978,14 @@ SEXP py_get_formals(PyObjectRef func, bool convert) {
   if (kw_only.is_null()) stop(py_fetch_error());
 
   Rcpp::Pairlist formals;
-  PyObject* param;
+  PyObjectPtr param;
   bool var_encountered = false;
-  while ((param = PyIter_Next(param_iter.get()))) {
-    PyObjectPtr param_name(PyObject_GetAttrString(param, "name"));
+  while (param.assign(PyIter_Next(param_iter.get())), !param.is_null()) {
```

Comment:
This still isn't quite right -- `assign()` doesn't call `Py_DecRef()` on a previously-held reference to a Python object, so I think there would still only be one `Py_DecRef()` being called at the end when the (single) `PyObjectPtr` is destructed.

One solution would be to construct a new `PyObjectPtr` for each iterated item, e.g.

```
while (true) {
    PyObjectPtr param(PyIter_Next(param_iter.get()))
    if (param == nullptr)
        break;
    < ... >
}
```

Reply:
That’s of course less pretty than what I was going for, but sure!

Alternatively: Is there a reason `assign()` doesn’t call `Py_DecRef()`? Seem like an oversight, unless I’m missing something.